linera-storage/src/lib.rs

```
@@ -332,6 +330,88 @@ pub struct ChainRuntimeContext<S> {
     user_services: Arc<DashMap<ApplicationId, UserServiceCode>>,
 }
 
+#[derive(Debug, Error)]
+pub enum ReadCertificateError {
+    MissingLiteCertificate(CryptoHash),
+    MissingConfirmedBlock(CryptoHash),
+    InconsistentHash(CryptoHash),
+    InconsistentEntries(CryptoHash),
+}
+
+impl std::fmt::Display for ReadCertificateError {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(f, "{:?}", self)
+    }
+}
+
+type ResultReadCertificate = Result<ConfirmedBlockCertificate, ReadCertificateError>;
+
+#[derive(Debug, Error)]
+pub struct ReadCertificatesError {
```

This looks complicated. The motivation for this PR is to simplify things.

---

It would be inexact to say that the motivation of this PR is to simplify things.

The motivation of this PR is to clean up things, to expose the right level of abstraction. You want simple? Put one error in `ViewError::NotFound` and the other in `ViewError::InconsistentEntries`. That is the current state of affairs, and we are not happy about it.

Is it possible to do another design? Surely:
* For example, in the query of certificates, we can take only the first incorrect one if it happens.
* Or, we can hide the reason for the hash failure behind an `IncorrectHashes(Vec<CryptoHash>)`.

Those choices are reasonable because we know where the approximation occurs. Also, it would show up directly in the error variant `IncorrectHash` or `IncorrectHashes`, expressing that something is wrong without exposing the exact reason for the error that needs to be dug.

That is an honest simplification. A dishonest simplification is to hide the error behind a generic `ViewError` or a variant of the same idea.